react/features/mobile/callkit/middleware.js

```diff
@@ -353,57 +328,31 @@ function _setCallKitSubscriptions({ getState }, next, action) {
 }
 
 /**
- * Notifies the feature callkit that the action {@link SET_VIDEO_MUTED} is being
- * dispatched within a specific redux {@code store}.
+ * Synchronize the muted state of tracks with CallKit.
  *
  * @param {Store} store - The redux store in which the specified {@code action}
  * is being dispatched.
  * @param {Dispatch} next - The redux {@code dispatch} function to dispatch the
  * specified {@code action} in the specified {@code store}.
- * @param {Action} action - The redux action {@code SET_VIDEO_MUTED} which is
- * being dispatched in the specified {@code store}.
+ * @param {Action} action - The redux action which is being dispatched in the
+ * specified {@code store}.
  * @private
  * @returns {*} The value returned by {@code next(action)}.
  */
-function _setVideoMuted({ getState }, next, action) {
-    const result = next(action);
-
-    const conference = getCurrentConference(getState);
-
-    if (conference && conference.callUUID) {
-        CallKit.updateCall(
-            conference.callUUID,
-            { hasVideo: !isVideoMutedByAudioOnly(getState) });
-    }
-
-    return result;
-}
-
-/**
- * Handles a track creation failure. This is relevant to us in the following
- * (corner) case: if the user never gave their permission to use the microphone
- * and try to unmute from the CallKit interface, this will fail, and we need to
- * sync back the CallKit button state.
- *
- * @param {Store} store - The redux store in which the specified {@code action}
- * is being dispatched.
- * @param {Dispatch} next - The redux {@code dispatch} function to dispatch the
- * specified {@code action} in the specified {@code store}.
- * @param {Action} action - The redux action {@code TRACK_CREARE_ERROR} which is
- * being dispatched in the specified {@code store}.
- * @private
- * @returns {*} The value returned by {@code next(action)}.
- */
-function _trackCreateError({ getState }, next, action) {
+function _syncTrackState({ getState }, next, action) {
     const result = next(action);
+    const { track } = action;
     const state = getState();
     const conference = getCurrentConference(state);
 
-    if (conference && conference.callUUID) {
+    if (track.local && conference && conference.callUUID) {
```

Comment:
`track.local` is always 'undefined' when muting/unmuting tracks, because TRACK_UPDATED does not carry full 'track' state from the reducer:
https://github.com/jitsi/jitsi-meet/blob/master/react/features/base/tracks/actions.js#L334
Adjusting it to `track.jitsiTrack.isLocal()` will trigger the update, but then it's possible to end up with an infinite loop when rapidly muting/unmuting as if the "break the loop" logic was no good (looking into that as well).

Reply:
It feels that the break the loop logic should be a check whether what CallKit wants is what base/media wants and emit setAudioMuted only if it does not match.

What goes to CallKit should be the reality though (base/tracks) - which seems to be fine currently when looking at _syncTrackState. Except for the call initiation part where the tracks may not exists yet and the conference will be trying to sync up with the intention while initializing, so I'm planning to adjust the logic in CallKit.startCall a little bit.

Reply:
I'm not sure what problem you are finding, can you please elaborate? The fact that `local` may be undefined? Hum, that would mean we run it too often, which indeed I've seen as a problem. However, upon adding that I couldn't repro it again 🤦‍♂️

Reply:
It's just that this condition is not executed on muting/unmuting local audio track, because it will never be true. I guess it works to some extent by a chance, because the update will be triggered when a local track is crated (either audio or video)